### PR TITLE
⚠️ Soft-deprecate `spacing` tokens and ship migration codemod (PR 1 of 3)

### PR DESCRIPTION
**Pull request by jandrade**
Adds `@deprecated` JSDoc to the `spacing` primitive token and its `VALID_*` type exports so editors surface deprecation warnings. Introduces `wb-codemod/transforms/migrate-spacing-to-sizing.ts` — a best-effort jscodeshift transform that:

- rewrites `spacing.<key>` → `sizing.<mapped>` in style objects, JSX attributes, and other plain references;
- rewrites arithmetic involving `spacing.*` into a `calc(...)` template literal composed from `sizing` tokens; and
- leaves a `// TODO(spacing-migration):` comment on patterns it can't safely auto-migrate (computed access, spread, `VALID_*` type imports).

No runtime or visual changes. Step 1 of the 3-PR rollout in `plan.md`; internal migration (PR 2) and `spacing` export removal (PR 3) follow.

Reviewers: #frontend-infra-web

Issue: WB-2314

## Test plan

- [x] `pnpm typecheck`, `pnpm jest`, `pnpm build:storybook` all clean (verified locally).
- [ ] In an editor (VS Code), open any consumer that imports `spacing` from `@khanacademy/wonder-blocks-tokens` (e.g. `packages/wonder-blocks-accordion/src/components/accordion-section-header.tsx`) and confirm the hover shows the `@deprecated` JSDoc and a strikethrough on `spacing` and `VALID_*` types.
- [ ] (Optional) Dry-run the codemod against a heavy consumer: `node wb-codemod/bin/cli.js -t migrate-spacing-to-sizing -d -p packages/wonder-blocks-accordion` and skim the output for sanity.

## Review plan

Please review this risky change:

1. ⚠️ [`migrate-spacing-to-sizing.ts`](https://github.com/Khan/wonder-blocks/pull/3031#discussion_r3174883985)

### Common patterns

**4 sites in 1 file:** `@deprecated` JSDoc on the four `spacing` public-API exports, all sharing one message that points at `sizing` and the codemod path.

**1 transform:** New jscodeshift codemod that walks each top-level `BinaryExpression` containing a `spacing.X` and rewrites the whole arithmetic tree as a `calc(...)` template literal. Numeric literals get `px` for `+/-` and stay unitless for `*//`. Example:

```ts
// Before
const INNER_BORDER_RADIUS = spacing.small_12 - 1;

// After
const INNER_BORDER_RADIUS = `calc(${sizing.size_120} - 1px)`;
```

🤖 Built using Claude Code #ai-generated

### CODE REVIEW

File: wb-codemod/transforms/migrate-spacing-to-sizing.ts [hunk 1]
```diff
@@ -0,0 +1,624 @@
+import {
```
Comment (jandrade) [inline wb-codemod/transforms/migrate-spacing-to-sizing.ts]: ⚠️ **1 of 1:** New ~620-LOC jscodeshift transform with arithmetic → `calc(...)` rewriting. Covered by 17 unit tests but worth a close read of `flattenBinary` / `buildCalcTemplate` before consumers run it on production code in PR 2.

**Comment by jandrade** [comment]
@claude review once

File: packages/wonder-blocks-tokens/src/tokens/spacing.ts [hunk 2]
```diff
@@ -16,6 +21,26 @@ export const spacing = {
     xxxLarge_64: 64,
 } as const;
 
+/**
+ * @deprecated The `spacing` token is deprecated; use `sizing` from
+ * `@khanacademy/wonder-blocks-tokens` instead. For APIs that need to constrain
+ * to a numeric set, prefer `number` (or `number | string` to also accept rem
+ * strings from `sizing`).
+ */
 export type VALID_PRIMARY_SPACING = 4 | 8 | 16 | 32 | 64;
```
Comment (beaesguerra) [inline packages/wonder-blocks-tokens/src/tokens/spacing.ts:30]: TIL we have a concept of primary/secondary spacing! Do you have more context/history around that? I'm curious! 😄

File: wb-codemod/transforms/migrate-spacing-to-sizing.ts [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+            return;
+        }
+
+        const key = (path.value.property as any).name as string;
+        const sizingKey = SPACING_TO_SIZING[key];
+        if (!sizingKey) {
+            return;
+        }
+
+        // `-spacing.X` (unary minus) on a numeric token used to produce a
```
Comment (beaesguerra) [inline wb-codemod/transforms/migrate-spacing-to-sizing.ts:529]: Nice ! I was going to ask if it handled negative spacing values. 

It would be also helpful to have a lint/stylelint rule to prevent `-${sizing.*}` usage going forward!

File: wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+import {spacing} from "@khanacademy/wonder-blocks-tokens";
+
+const Spacer = () => <Strut size={spacing.xxxSmall_4} />;
+`,
+            `
+import * as React from "react";
+import {Strut} from "@khanacademy/wonder-blocks-layout";
+import {sizing} from "@khanacademy/wonder-blocks-tokens";
+
+const Spacer = () => <Strut size={sizing.size_040} />;
```
Comment (beaesguerra) [inline wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts]: One difference between `spacing` and `sizing` tokens is that `spacing` is a number and `sizing` is a string. 

When `spacing` tokens are converted to `sizing` tokens, there will be type errors where `Strut` is used with these tokens since `size` expects a number! Would part of this work also include shifting away from `Strut`? Or allowing string values for the `size` prop?

File: wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+            `
+import {spacing} from "@khanacademy/wonder-blocks-tokens";
+const styles = {padding: \`10px \${spacing.medium_16}px\`};
+`,
+            `
+import {spacing} from "@khanacademy/wonder-blocks-tokens";
+// TODO(spacing-migration): manual review needed — \`spacing\` reference could not be auto-migrated.
+const styles = {padding: \`10px \${spacing.medium_16}px\`};
+`,
+            "should bail when spacing is followed by a CSS unit suffix in a template literal",
```
Comment (beaesguerra) [inline wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts:715]: Another lint rule idea is no CSS unit suffix for sizing tokens going forward!

File: wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+            `
+import * as React from "react";
+import {Button} from "@khanacademy/wonder-blocks-button";
+
+export const App = () => <Button>Click me</Button>;
+`,
+            "should produce identical output for files with no spacing references",
+        );
+    });
+});
```
Comment (beaesguerra) [inline wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts:803]: Great set of tests!

**Review by beaesguerra — COMMENTED** [review]
Overall looks good to me, I left some questions! I see the Claude code review also flagged some things, I'll re-review again once those are addressed!

File: packages/wonder-blocks-tokens/src/tokens/spacing.ts (hunk 2 above)
Comment (jandrade) [reply packages/wonder-blocks-tokens/src/tokens/spacing.ts:30]: TBH that existed even before I joined, and I've never used that 😄. I checked in the codebase and it is never used (including other repos), and I was inclined to remove it, but decided to keep it around to simplify the scope of this PR.

File: wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts (hunk 4 above)
Comment (jandrade) [reply wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts]: oh nice catch! the goal is to shift away from `Strut`. This change is happening in the second PR. For now I'm advocating for removing `Strut`, but might change to allowing string values if the `frontend` migration turns complex with some edge cases.

For now I'm going to get rid of this line to avoid giving the wrong signal.

File: wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts (hunk 5 above)
Comment (jandrade) [reply wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts:715]: ohh that's a great call!

File: wb-codemod/transforms/migrate-spacing-to-sizing.ts [hunk 7]
```diff
@@ -0,0 +1,760 @@
...
+        if (isInsideTemplateLiteralExpression(path)) {
+            return;
+        }
+
+        const parts = flattenBinary(j, path.value, spacingBinding, null);
+        if (!parts) {
+            return;
+        }
+
+        const template = buildCalcTemplate(j, parts, sizingBinding);
```
[the remark this answers is not in the document]
Comment (jandrade) [reply wb-codemod/transforms/migrate-spacing-to-sizing.ts]: Fixed in https://github.com/Khan/wonder-blocks/pull/3031/changes/2d03152beeaae8e7f1927f60815a6b68bede6fd8

File: wb-codemod/transforms/migrate-spacing-to-sizing.ts [hunk 8]
```diff
@@ -0,0 +1,760 @@
...
+        node.object?.type === "Identifier" &&
+        node.object.name === spacingBinding &&
+        node.property?.type === "Identifier" &&
+        SPACING_TO_SIZING[node.property.name] !== undefined
+    );
+}
+
+function containsSpacingReference(
+    j: JSCodeshift,
+    path: ASTPath<BinaryExpression>,
```
[the remark this answers is not in the document]
Comment (jandrade) [reply wb-codemod/transforms/migrate-spacing-to-sizing.ts:861]: Fixed in https://github.com/Khan/wonder-blocks/pull/3031/changes/313bf0648c9983959b7c4a01e92d0b5d1f02fb92

File: wb-codemod/transforms/migrate-spacing-to-sizing.ts [hunk 9]
```diff
@@ -0,0 +1,760 @@
...
+        // TODO comment.
+        if (isInsideTemplateLiteralExpression(path)) {
+            return;
+        }
+
+        const parts = flattenBinary(j, path.value, spacingBinding, null);
+        if (!parts) {
+            return;
+        }
+
```
[the remark this answers is not in the document]
Comment (jandrade) [reply wb-codemod/transforms/migrate-spacing-to-sizing.ts:472]: Fixed in https://github.com/Khan/wonder-blocks/pull/3031/changes/7bbfcc49882a651082ea0a68783f62d851d8ae51

File: wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts (hunk 5 above)
Comment (beaesguerra) [reply wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts:715]: I added a section for [CSS related lint / stylelint rules](https://khanacademy.atlassian.net/wiki/spaces/WB/pages/4617732202/WB+lint+rules+exploration#CSS-Related-lint%2Fstylelint-rules) and included some of the edge cases around using sizing tokens. Feel free to add to it! 

cc: @marcysutton also since it's related to lint rules (These are probably lower priority compared to the other rules we've prioritized!)

File: wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts (hunk 4 above)
Comment (jandrade) [reply wb-codemod/transforms/__tests__/migrate-spacing-to-sizing.test.ts]: Fixed in https://github.com/Khan/wonder-blocks/pull/3031/changes/e5dae267edaebd808018ff6510d6faa3106dd58f

But then switched to a different strategy: If there's a complex `Strut` case, then the codemod will add a TODO comment to do a manual fix, otherwise it will migrate to use `gap: sizing.size_xxx`: https://github.com/Khan/wonder-blocks/pull/3031/changes/e5dae267edaebd808018ff6510d6faa3106dd58f

**Review by beaesguerra — APPROVED** [review]
Looks good to me!
